gc.c

```diff
@@ -6080,11 +6081,18 @@ invalidate_moved_plane(rb_objspace_t *objspace, struct heap_page *page, uintptr_
                     CLEAR_IN_BITMAP(GET_HEAP_PINNED_BITS(forwarding_object), forwarding_object);
 
                     object = rb_gc_location(forwarding_object);
+                    if (RB_TYPE_P(object, T_OBJECT)) {
+                        old_shape_id = rb_shape_get_shape_id(forwarding_object);
+                    }
 
                     gc_move(objspace, object, forwarding_object, GET_HEAP_PAGE(object)->slot_size, page->slot_size);
                     /* forwarding_object is now our actual object, and "object"
                      * is the free slot for the original page */
 
+                    if (RB_TYPE_P(forwarding_object, T_OBJECT)) {
```

Comment:
No problem. It originates from here: https://github.com/Shopify/ruby/blob/mvh-object-movement-shapes/gc.c#L8504

If we find a new destination shape in the new size pool, we set it during the object movement set of compaction, this avoids us having to do any work in update references

However we also need to be able to return to the original shape if for some reason the page gets invalidated - so we store the original shape id in the T_MOVED object itself, so that when it becomes invalid and moves back we can restore the original shape.